tests/integration_tests/test_kernel_commandline_match.py

```diff
@@ -14,10 +14,9 @@ def override_kernel_cmdline(ds_str: str, c: IntegrationInstance) -> str:
     Ironic, for example, it will succeed.
     """
     client = c
+    client.execute("sed --in-place /GRUB_CMDLINE_LINUX=/d /etc/default/grub")
     client.execute(
-        "sed --in-place "
-        "'s/^.*GRUB_CMDLINE_LINUX=.*$/GRUB_CMDLINE_LINUX=\"" + ds_str + '"/g'
-        "' /etc/default/grub"
+        "echo \"GRUB_CMDLINE_LINUX='" + ds_str + "'\" >> /etc/default/grub"
```

Comment:
I don't think we've quite gotten here as the this test still passed against ppa:cloud-init-dev/daily. On looking at the test instance, defaults which are getting written to /etc/default/grub still look like they are set 'correctly' but /proc/cmdline across boot also shows the same truncation before the semi-colon. This means we are still testing the simple-case where the kernel cmdline param is only `ds=nocloud-net`   not `ds=nocloud-net;s=http://something-else`
```
root@cloudinit-0410-145717kkrvkg7w:~# grep ds= /etc/default/grub
GRUB_CMDLINE_LINUX='ds=nocloud;s=http://my-url/'
root@cloudinit-0410-145717kkrvkg7w:~# cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.4.0-1086-kvm root=PARTUUID=bce261e6-76e0-4f07-a5a1-f5ed57bab053 ro ds=nocloud
```


What we need is the value in /etc/default/grub to be wrapped in both double and single quotes grub shell doesn't ignore anything beyond the semi-colon:
```
GRUB_CMDLINE_LINUX="'dsnocloud-net;s=http://my-url/'"
```


I just tested with the following patch and can now see expected failures on daily. 
```diff
diff --git a/tests/integration_tests/test_kernel_commandline_match.py b/tests/integration_tests/test_kernel_commandline_match.py
index dbef11020..e3f5e582e 100644
--- a/tests/integration_tests/test_kernel_commandline_match.py
+++ b/tests/integration_tests/test_kernel_commandline_match.py
@@ -15,8 +15,9 @@ def override_kernel_cmdline(ds_str: str, c: IntegrationInstance) -> str:
     """
     client = c
     client.execute("sed --in-place /GRUB_CMDLINE_LINUX=/d /etc/default/grub")
+    ds_str=f"'{ds_str}'"
     client.execute(
-        "echo \"GRUB_CMDLINE_LINUX='" + ds_str + "'\" >> /etc/default/grub"
+        f"echo \"GRUB_CMDLINE_LINUX={ds_str}\" >> /etc/default/grub"
     )
 
     # We should probably include non-systemd distros at some point. This should
```

Reply:
Looks like the proposed fix here only adds single quotes, not the double quotes that are also required. I'll cook up a fix.

Reply:
@blackboxsw should be fixed now:
```
root@cloudinit-0410-190811ms45vji5:~# cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.15.0-1030-kvm root=PARTUUID=52746263-2501-4e3e-b990-e29924f87bd0 ro ds=nocloud;s=http://my-url/ console=tty1 console=ttyS0 panic=-1
```

See the latest commit which passes the correct commandline all the way through the stack to `/boot/efi/EFI/ubuntu/grub.cfg`

Reply:
@holmanb +1 on this just confirmed expected failure against ppa:cloud-init-dev/daily and the fix seen once we have your locally built deb from this branch. Also seeing the correct breadcrumbs showing we've properly trucanted the /proc/cmdline values when searching for matching datasources.
```
2023-04-11 02:08:40,372 - __init__.py[DEBUG]: Looking for data source in: ['nocloud', 'None'], via packages ['', 'cloudinit.sources'] that matches dependencies ['FILESYSTEM']
```